origin-discovery/src/apollo/db.js

```diff
@@ -77,10 +92,42 @@ async function getListingsBySeller (sellerAddress) {
 /**
  * Queries DB for a listing.
  * @param listingId
+ * @param {Object} blockInfo - Optional max blockNumber and logIndex values (inclusive).
+ *   This can be used to get the state of a listing at a given point in history.
  * @return {Promise<Object|null>}
  */
-async function getListing (listingId) {
-  const row = await db.Listing.findByPk(listingId)
+async function getListing (listingId, blockInfo = null) {
+  let row
+  if (blockInfo) {
+    // Build a query that looks like:
+    // SELECT * FROM listing WHERE id=listingId AND
+    //      (block_number < blockInfo.blockNumber OR
+    //      (block_number = blockInfo.blockNumber AND log_index <= blockInfo.logIndex))
+    // ORDER BY block_number DESC, log_index DESC
+    // LIMIT 1
+    row = await db.Listing.findOne({
```

Comment:
Nice! This solution covers the corner case you have described in the comment previously. I would only add a comment explaining this query with the example you used last time to help me understand the problem.

Reply:
Good idea. Added example to comments.